jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletMultiPartFormData.java

```diff
@@ -51,15 +54,88 @@
  */
 public class ServletMultiPartFormData
 {
+    /**
+     * Get {@code multipart/form-data} {@link ServletMultiPartFormData.Parts} from a {@link ServletRequest}, caching the
+     * results in the request {@link ServletRequest#getAttribute(String) Attributes}.  If not already available,
+     * the {@code Parts} are read and parsed, blocking if necessary.
+     * <p>
+     * Calls to {@code onParts} and {@code getParts} methods are idempotent, and
+     * can be called multiple times, with subsequent calls returning the results of the first call.
+     * @param servletRequest A servlet request
+     * @return the parts
+     */
+    static Parts getParts(ServletRequest servletRequest)
+    {
+        CompletableFuture<Parts> futureParts = from(servletRequest, org.eclipse.jetty.util.thread.Invocable.InvocationType.NON_BLOCKING);
+        return futureParts.join();
+    }
+
+    /**
+     * Asynchronously get {@code multipart/form-data} {@link ServletMultiPartFormData.Parts} from a {@link ServletRequest},
+     * caching the results in the request {@link ServletRequest#getAttribute(String) Attributes}.  If not already available,
+     * the {@code Parts} are read and parsed.
+     * <p>
+     * Calls to {@code onParts} and {@code getParts} methods are idempotent, and
+     * can be called multiple times, with subsequent calls returning the results of the first call.
+     * @param servletRequest A servlet request
+     * @param contentType The contentType, passed as an optimization as it has likely already been retrieved.
+     * @param immediate The action to take if the Parts are available immediately (from within the scope of the call to this method).
+     * @param future The action to take when the Parts are available, if they are not available immediately.  The {@link org.eclipse.jetty.util.thread.Invocable.InvocationType}
+     *               of this parameter will be used as the type for any implementation calls to {@link Content.Source#demand(Runnable)}.
+     */
+    static void onParts(ServletRequest servletRequest, String contentType, Promise<Parts> immediate, Promise.Invocable<Parts> future)
```

Comment:
Not 2 promises for the same result.

Reply:
See discussion in FormFields for resolution